### ISSUE TITLE
Train_test_split assist v1 - create warnings

### ISSUE THREAD

**Issue by MarieS-WiMLDS** [probabl-ai/skore#492]
As a data scientist, I want to be guided in the choice of the arguments in the scikit-learn train_test_split function, without having too many warnings to avoid being over my cognitive budget charge (let's say: 2 warnings max).  

### Warnings

Here is a list of warnings to output to the users when they call skore.train_test_split, based on [sklearn.model_selection.train_test_split](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html).  

Note: warnings have an advice (or message), that can sometimes refer to the documentation (of the same library version).

W1:
- Advice: It seems that you have a classification problem with a high class imbalance. In this case, using `train_test_split` may not be a good idea because of high variability in the scores obtained on test set. To tackle this challenge we suggest to use the skore's function cross_validate _{put a link to the doc when it's ready ;)}_.
- Trigger: 
  - AND
    - `stratify` is not set
    - We detect a target with one of the following hypothesis (in this order): 
      - A variable called `y` is given to the the function in array parameters
      - A variable called `target` is given to the the function in array parameters
      - the user passes more than 2 arrays, and the last one is 1-dimensional, we assume it's y
    - The target is detected as a "classification" ML task
    - Given the number of examples from the most important class n_max and the number of examples from the least important class n_min, `n_max/n_min > 3` must be verified

W1 bis:
- Advice: It seems that you have a classification problem with at least one class with less than 100 examples in the test set. In this case, using `train_test_split` may not be a good idea because of high variability in the scores obtained on test set. We suggest three options to tackle this challenge: you can increase test_size, collect more data, or use the skore's function cross_validate _{put a link to the doc when it's ready ;)}_.
- Trigger: 
  - AND
    - `stratify` is not set
    - We detect a target with one of the following hypothesis (in this order): 
      - A variable called `y` is given to the the function in array parameters
      - A variable called `target` is given to the the function in array parameters
      - the user passes more than 2 arrays, and the last one is 1-dimensional, we assume it's y
    - The target is detected as a "classification" ML task
    - In the test set I have a class with less than 100 examples.

W2: 
- Advice: We recommend against using `stratify` if you have a problem where your events occur over time (even if they are independent). Using `stratify` will result in inflated model performance evaluation because natural drift of the target will not be taken into account.
- Trigger:
  - `stratify` is not empty
 
W3:
- Advice: We recommend setting the parameter `random_state`. This will ensure the reproducibility of your work.
- Trigger:
  - AND
    - `shuffle` is set to True.
    - No random state in the parameters.

W4: 
- Advice: We detected a time-based column ({output the dataframe name}, column {output col name if available or index otherwise}) in your data. We recommend using [TimeSeriesSplit](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.TimeSeriesSplit.html) instead of `train_test_split`. Otherwise you might train on future data to predict the past, or get inflated model performance evaluation because natural drift will not be taken into account.
- Trigger:
  - We detect a time column somewhere in one of the arrays. (I don't think it will be possible when we receive numpy arrays, and it's ok, it's best effort)

W5:
- Advice: We detected the parameter `shuffle` is set to `True` either explicitly or from its default value. In case of time-ordered events (even if they are independent), this will result in inflated model performance evaluation because natural drift will not be taken into account. We recommend setting the `shuffle` parameter to `False` in order to ensure the evaluation process is really representative of your production release process.
- Trigger: 
  - The effective value of the `shuffle` parameter is `True`, or it's empty

List of priority for the warnings: W4 before W1 before W5 before W3.

### Output

1. We would like the user to be able to retrieve their usual output of train_test_split from sklearn. 
2. We would like the warnings to be picked up by Jupyter or VS Code ipython plugin. It looks like can be simply implemented by issuing a warning from the `warnings` standard Python library.

### Additional note

In these warnings, we make the hypothesis that the dataset represents events occurring over time, even though they are independent. We suppose that it's the case most of the time in real-world application. There are some cases where it's not, for instance when trying to understand how a molecule interact with a virus. 

## TODO

- [x] #668
- [x] https://github.com/probabl-ai/skore/issues/683
- [x] https://github.com/probabl-ai/skore/issues/684

**Comment by MarieS-WiMLDS** [comment]
suggestion: instead of warning, find a term more user-friendly. dataiku uses the term "ML-diagnostic".

**Comment by MarieS-WiMLDS** [comment]
List of subissues of this epic:  
- [ ] technical design
- [ ] implement tech design

No front is required for now. We definitely want to integrate these warnings in skore UI at some point, and/or in Ensembl. It requires first some UX design about how we want to do this.

**Comment by augustebaum** [comment]
For W4, do you mean we should detect when a column has dtype "datetime"? Or also if it's a string that looks like a datetime?

**Comment by augustebaum** [comment]
For W1, do you mean "at least one class has fewer than 100 examples" or "all classes have fewer than 100 examples"? the former seems more appropriate when we talk about class imbalance right?

**Comment by MarieS-WiMLDS** [comment]
(I wish I could answer comments!!)  

> For W4, do you mean we should detect when a column has dtype "datetime"? Or also if it's a string that looks like a datetime?  

Whenever you find a solution to detect a datetime! If you think that it's possible in a reliable way to detect if a string looks like a datetime even without the dtype, let's go! The tech answer is up to you :)

> For W1, do you mean "at least one class has fewer than 100 examples" or "all classes have fewer than 100 examples"? the former seems more appropriate when we talk about class imbalance right?  

You're right, it's the smallest class that should have more than 100 examples nonetheless. I'll edit the issue description to make it clearer.

**Comment by augustebaum** [comment]
For W2 I don't understand; do we want users to use `stratify`? Or are we saying "if you choose to use `stratify`, here is what you need to know"? At a glance, I don't understand the advice that is given

**Comment by emmatysinger** [comment]
W1 –  It seems that you have a classification problem with a high **class** imbalance, the test set has less than 100 examples of each class. **In this case**, using train_test_split **may** not be a good idea because of high variability in the scores obtained on test set. We suggest three options to tackle this challenge: you can increase test_size, collect more data, or use the skore's function cross_validate {put a link to the doc when it's ready ;)}.

W2 – We advise **against (?)** using stratify if you have a problem where your events occur over time (even if they are independent). **Using `stratify`** will result in optimistic evaluation results because natural drift of the target will not be taken into account.
Q: not sure I understand the use of this warning. You want to recommend using stratify if the trigger is the user is already using stratify?

W6 – We detected the parameter shuffle is set to True either explicitly or from its default value. In case of time-ordered events (even if they are independent), this will result in optimistic evaluation results because natural drift will not be taken into account. We **recommend** setting the shuffle parameter _explicitly (remove explicitly)_  to False in order to ensure the evaluation process is really representative of your production release process.

Notes: 

- I would say **recommend** instead of **advise**
- I would say **inflated/biased model performance** instead of **optimistic result** for W2, W4 and W6

**Comment by MarieS-WiMLDS** [comment]
Thanks Emma for your feedback!  

Using recommend or advise was one of our questions at some point. We chose advise because it seemed less mandatory (at least it's the case in french). Why would you use recommend, because it's more frequent, because it matches best the overall tone...?

Love your suggestion on inflated model performance, much clearer!

And indeed for warning 2, just like Auguste pointed, a word is missing! 

I'm updating the advice texts accordingly.

**Comment by MarieS-WiMLDS** [comment]
ok, let's switch to recommend then :) !

**Comment by augustebaum** [comment]
From W1:
> Given the number of examples from the most important class n_max and the number of examples from the least important class n_min, n_max/n_min > 3 must be verified

Is that the number of examples in the test set?

**Comment by MarieS-WiMLDS** [comment]
> Does this refer to y_test?

It refers to the target. Only, in [train_test_split](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html), you can pass as many arrays as you want, in any order you want. We would rely on the name of the array to guess that it's the target.

**Comment by augustebaum** [comment]
Yeah so I meant to tell you about this: the arrays in train_test_split are passed *positionally*, not by keyword. That means we cannot know for sure which array is the target, we can only make guesses (e.g. the second array is often the target, and the target often is a 1-column array).

**Comment by thomass-dev** [comment]
@MarieS-WiMLDS , for W1, you said raise a warning if 
> In the test set I have a class with less than 100 examples.

But, dealing with small dataset, i can have 3 homogeneous classes with each less than 100 examples.
Why in that case do i have a warning?

**Comment by MarieS-WiMLDS** [comment]
@thomass-dev, great remark. I'll split W1 into these two cases: imbalance and low amount of data.

**Comment by MarieS-WiMLDS** [comment]
@augustebaum: instead of using a positional hypothesis, I'm suggesting to use the name of the variable.  
About your hypothesis `if the user passes more than 2 arrays, can we assume the last one is y` -> i like it, and i'm pretty sure we'll be right 99.9% of the time. I think we can add it, even if it means removing it after a while if we have too many feedbacks indicating the guess wrong.

**Comment by MarieS-WiMLDS** [comment]
After discussion: we agreed to ask the user X, y, *arrays for the arrays to split, to ensure that we always know where y is. Therefore we never output a warning while we shouldn't, and we never don't output a warning while we should. Also, for most people it's still transparent from scikit-learn as most people call the scikit-learn function with `train_test_split(X,y)`.

**Comment by augustebaum** [comment]
After discussion with @MarieS-WiMLDS it looks like we might want to fuse warnings W5 and W6 together. Essentially, if the user does not set `shuffle`, then the default is to set it to True. `shuffle=True` is bad because doing so means assuming that the data are independent and identically distributed, and production data often does not satisfy this property (for example, there could be a distribution shift between older and newer data).

Thus, here is the new design:
- The default value for `shuffle` in `skore.train_test_split` becomes `None`. That way, we can tell if the user expects the default value or if they explicitly set the value.
	- If left at None, we pass `shuffle=True` down to sklearn, and W6 will trigger
	- If set to True, W6 will trigger as originally designed
	- If set to False, there will be no warning

**Comment by MarieS-WiMLDS** [comment]
This epic is done :white_check_mark:  !
